### ISSUE TITLE
Alpine musl libc build issue

### ISSUE THREAD

**Issue by jamespedwards42** [zcash/zcash#1558]
Because of the build flag -Werror in [zcutil/build.sh](https://github.com/zcash/zcash/blob/master/zcutil/build.sh#L40) which causes warnings to be treated as errors, the build fails on alpine linux because:

* `#warning redirecting incorrect #include <sys/fcntl.h> to <fcntl.h>`
* `#warning redirecting incorrect #include <sys/poll.h> to <poll.h>`

I'm not sure what the best way to resolve this issue is, just hoping it might be obvious to someone else.

I've started a docker file at [github.com/jamespedwards42/alpine-zcash/blob/master/Dockerfile](https://github.com/jamespedwards42/alpine-zcash/blob/master/Dockerfile) if anyone wants to give it a go.

tail of the failed build logs: [pastebin.com/raw/0irE64dt](http://pastebin.com/raw/0irE64dt)

**Comment by daira** [comment]
Testing a fix now.

**Comment by daira** [comment]
Please test using a new checkout with this ``git clone`` command (instead of the ``git clone``/``checkout`` in the Beta Guide):
```
git clone -b 1558.alpine-linux-fixes https://github.com/daira/zcash.git
```

**Comment by jamespedwards42** [comment]
One more straggler, but it looks like that is going to get it.

build logs: [pastebin.com/BV09fHTc](http://pastebin.com/BV09fHTc)

**Comment by daira** [comment]
I'm a bit confused because the fix should already have caught that one. What is at line 61 of ``/usr/src/zcash/depends/x86_64-unknown-linux-gnu//include/boost/asio/detail/socket_types.hpp``? Also, did you perform the clone in a new directory? (Just switching to that branch in an existing checkout may not work.)

**Comment by jamespedwards42** [comment]
It is from a docker build, so it is always a completely fresh filesystem.

**Comment by jamespedwards42** [comment]
Heres is what I ran:

```dockerfile
RUN mkdir -p /usr/src \
  && cd /usr/src \
  && git clone -b 1558.alpine-linux-fixes https://github.com/daira/zcash.git \
  && cd zcash \ 
  && /bin/bash ./zcutil/fetch-params.sh

RUN cd /usr/src/zcash \
  && /bin/bash ./zcutil/build.sh -j4
```

**Comment by jamespedwards42** [comment]
Running again without any previous layer caching just to double check.

**Comment by jamespedwards42** [comment]
no luck, same error.

**Comment by daira** [comment]
And what is at ``/usr/src/zcash/depends/x86_64-unknown-linux-gnu/include/boost/asio/detail/socket_types.hpp`` after the failed build?

If it still says ``#include <sys/poll.h>``, can you link the full build log?

**Comment by jamespedwards42** [comment]
[Unfortunately, I cannot mount](https://github.com/docker/docker/issues/17745) to my local filesystem to see the files inside the docker build image.  I will run through the build on an Ubuntu server sometime tonight (+12-16 hours) to check it out.

Here is a complete build log: [hub.docker.com/r/jamespedwards42/alpine-zcash](https://hub.docker.com/r/jamespedwards42/alpine-zcash/builds/buelzpv82ueizaxjkikygcu/)

**Comment by daira** [comment]
The build copies ``include_poll.patch`` to the right place, and then doesn't apply it, even though it does apply  ``deprecated_auto_ptr.patch``. (Search for ``Preprocessing boost`` in the build log.) I'm flummoxed. Can anyone see why this isn't working?

**Comment by daira** [comment]
@jamespedwards42 Please try again now. (Not sure this will work; I changed the ``include_poll.patch`` format to match the other patch, but I don't see why it would make a difference.)

**Comment by str4d** [comment]
@daira oh, you need to manually add the patch commands. See the existing patch command in `define $(package)_preprocess_cmds`.

**Comment by jamespedwards42** [comment]
@daira here is a build triggered 29 minutes ago:

https://hub.docker.com/r/jamespedwards42/alpine-zcash/builds/b9gfk92fepyuf4hsggyvjck/

**Comment by daira** [comment]
Should be fixed now, please try again.

**Comment by jamespedwards42** [comment]
Thank you @daira!  The build succeeded, I'm working through the missing runtime dependencies that need to be installed.  Hopefully, I will have this wrapped up after tomorrow night.  Once everything is stable I will have it default to the latest release tag instead of your repo/branch.  Started a readme for how to use it here: [github.com/jamespedwards42/alpine-zcash](https://github.com/jamespedwards42/alpine-zcash)

**Comment by daira** [comment]
Assuming fixed by #1559. Please reopen if that's not the case.
